### PR TITLE
Adds 2 factor backend code

### PR DESCRIPTION
**Pull request by thatarchguy**
Ripped from #165 because of decision to separate the merge into frontend and backend pieces [https://github.com/StackFocus/PostMaster/pull/165#issuecomment-254382905](https://github.com/StackFocus/PostMaster/pull/165#issuecomment-254382905)

### CODE REVIEW

File: postmaster/apiv1/admins.py [hunk 1]
```diff
@@ -147,3 +149,115 @@ def unlock_admin(admin_id):
...
+            except Exception as e:
+                db.session.rollback()
+                json_logger(
+                    'error', current_user.username,
+                    'The following error occurred in twofactor_disable: {0}'.format(str(e)))
+                raise GenericError('The administrator could not be updated')
+            return dict(enabled=admin.otp_active)
+        elif status.lower() == "true":
+            raise GenericError("Cannot enable 2 factor from this route - see docs")
+    raise GenericError("An Invalid parameter was supplied")
```
Comment (mprahl) [inline postmaster/apiv1/admins.py]: Lower case I in "invalid"

File: postmaster/apiv1/admins.py [hunk 2]
```diff
@@ -147,3 +149,115 @@ def unlock_admin(admin_id):
...
+                db.session.commit()
+            except ValidationError as e:
+                raise e
+            except Exception as e:
+                db.session.rollback()
+                json_logger(
+                    'error', current_user.username,
+                    'The following error occurred in twofactor_disable: {0}'.format(str(e)))
+                raise GenericError('The administrator could not be updated')
+            return dict(enabled=admin.otp_active)
```
Comment (mprahl) [inline postmaster/apiv1/admins.py]: I don't think we should return anything in this case. Just an HTTP 200.

File: postmaster/apiv1/admins.py [hunk 3]
```diff
@@ -147,3 +149,115 @@ def unlock_admin(admin_id):
...
+                    json_logger('audit', current_user.username, audit_message)
+                except ValidationError as e:
+                    raise e
+                except Exception as e:
+                    db.session.rollback()
+                    json_logger(
+                        'error', current_user.username,
+                        'The following error occurred in verify_qrcode: {0}'.format(str(e)))
+                    raise GenericError('The administrator could not be updated')
+            return dict(status="Success")
```
Comment (mprahl) [inline postmaster/apiv1/admins.py]: I don't think we should be returning this here. We don't anywhere else in the app. A simple HTTP 200 should do here with no JSON.

File: postmaster/forms.py [hunk 4]
```diff
@@ -5,7 +5,7 @@
 """
 
 from flask_wtf import Form
-from wtforms import StringField, PasswordField, SelectField
+from wtforms import StringField, PasswordField, SelectField, IntegerField, validators
```
Comment (mprahl) [inline postmaster/forms.py]: Instead of importing `validators`, import `Optional` from it on the next line. Alternatively, you can remove the line below and do `validators.DataRequired()` on line 17, 18, and 20.

File: postmaster/forms.py [hunk 5]
```diff
@@ -16,6 +16,7 @@ class LoginForm(Form):
     """
     username = StringField(label='Username', validators=[DataRequired()])
     password = PasswordField(label='Password', validators=[DataRequired(), validate_wtforms_password])
+    two_factor = IntegerField(label='2 Factor', validators=(validators.Optional(),))
```
Comment (mprahl) [inline postmaster/forms.py]: Can the label be "Two Factor Code"

Comment (mprahl) [inline postmaster/forms.py]: I'm also thinking that instead of this, we should pop up a modal if two factor authentication is enabled. This is fine for now though.

File: postmaster/models.py [hunk 6]
```diff
@@ -262,7 +266,8 @@ def to_json(self):
         """
         return {'id': self.id, 'name': self.name, 'username': self.username, 'failed_attempts': self.failed_attempts,
                 'last_failed_date': self.last_failed_date, 'unlock_date': self.unlock_date,
-                'locked': (self.unlock_date is not None and self.unlock_date > datetime.utcnow())}
+                'locked': (self.unlock_date is not None and self.unlock_date > datetime.utcnow()),
+                'twoFactor': self.otp_active}
```
Comment (mprahl) [inline postmaster/models.py]: Can you name this "two_factor"?

File: postmaster/static/css/styles.css [hunk 7]
```diff
@@ -242,6 +242,10 @@ ul.pagination li.disabled a:hover {
 a.adminLocked {
     color: #e74c3c;
 }
+
```
Comment (mprahl) [inline postmaster/static/css/styles.css]: Can you remove this from the PR?

File: postmaster/static/css/styles.css [hunk 8]
```diff
@@ -289,36 +293,36 @@ a.adminLocked {
...
 
     /* No More Tables */
-    #dynamicTable, 
-    #dynamicTable thead, 
-    #dynamicTable tbody, 
-    #dynamicTable th, 
-    #dynamicTable td, 
-    #dynamicTable tr { 
-        display: block; 
+    #dynamicTable,
```
Comment (mprahl) [inline postmaster/static/css/styles.css]: Can you make the rest of this file a separate commit called `Clean up whitespace`?

File: postmaster/utils.py [hunk 9]
```diff
@@ -255,44 +255,41 @@ def validate_wtforms_password(form, field):
...
                         friendly_username = ad_object.get_loggedin_user()
                         display_name = ad_object.get_loggedin_user_display_name()
-
                         if not models.Admins.query.filter_by(username=friendly_username, source='ldap').first():
                             add_ldap_user_to_db(friendly_username, display_name)
-
                         admin = models.Admins.query.filter_by(username=friendly_username, source='ldap').first()
+                        if admin.otp_active:
+                            if not admin.verify_totp(two_factor):
+                                raise WtfStopValidation('2 Factor token was incorrect')
```
Comment (mprahl) [inline postmaster/utils.py]: Change this to `raise WtfStopValidation('The two factor authentication code is incorrect')`

File: postmaster/utils.py [hunk 10]
```diff
@@ -255,44 +255,41 @@ def validate_wtforms_password(form, field):
...
                 admin = models.Admins.query.filter_by(username=username, source='local').first()
-
                 if admin:
-
                     if admin.is_unlocked():
-
                         if bcrypt.check_password_hash(admin.password, password):
+                            if admin.otp_active:
+                                if not admin.verify_totp(two_factor):
+                                    raise WtfStopValidation('2 Factor token was incorrect')
```
Comment (mprahl) [inline postmaster/utils.py]: Change this to `raise WtfStopValidation('The two factor authentication code is incorrect')`

File: postmaster/utils.py [hunk 11]
```diff
@@ -255,44 +255,41 @@ def validate_wtforms_password(form, field):
...
         try:
             if form.auth_source.data == 'PostMaster User':
                 admin = models.Admins.query.filter_by(username=username, source='local').first()
-
                 if admin:
-
                     if admin.is_unlocked():
-
                         if bcrypt.check_password_hash(admin.password, password):
+                            if admin.otp_active:
```
Comment (mprahl) [inline postmaster/utils.py]: You can consolidate these two lines to:
`if admin.otp_active and not admin.verify_totp(two_factor):`

File: postmaster/utils.py [hunk 12]
```diff
@@ -255,44 +255,41 @@ def validate_wtforms_password(form, field):
...
-
                     if ad_object.check_group_membership():
                         friendly_username = ad_object.get_loggedin_user()
                         display_name = ad_object.get_loggedin_user_display_name()
-
                         if not models.Admins.query.filter_by(username=friendly_username, source='ldap').first():
                             add_ldap_user_to_db(friendly_username, display_name)
-
                         admin = models.Admins.query.filter_by(username=friendly_username, source='ldap').first()
+                        if admin.otp_active:
```
Comment (mprahl) [inline postmaster/utils.py]: You can consolidate these two lines to:
`if admin.otp_active and not admin.verify_totp(two_factor):`

File: postmaster/apiv1/admins.py [hunk 13]
```diff
@@ -147,3 +149,115 @@ def unlock_admin(admin_id):
     admin = Admins.query.get_or_404(admin_id)
     clear_lockout_fields_on_user(admin.username)
     return {}, 200
+
+
+@apiv1.route('/admins/<int:admin_id>/2factor', methods=['GET'])
```
Comment (mprahl) [inline postmaster/apiv1/admins.py]: Can we change the route to "twofactor"?

File: postmaster/apiv1/admins.py [hunk 14]
```diff
@@ -147,3 +149,115 @@ def unlock_admin(admin_id):
...
+    """ Returns if 2 factor is enabled or not
+
+        This information is in the main user route,
+        but I added it here as a stub for the URI.
+    """
+    admin = Admins.query.get_or_404(admin_id)
+    return dict(enabled=admin.otp_active)
+
+
+@apiv1.route('/admins/<int:admin_id>/2factor', methods=['PUT'])
```
Comment (mprahl) [inline postmaster/apiv1/admins.py]: Can we change the route to "twofactor"?

File: postmaster/apiv1/admins.py [hunk 15]
```diff
@@ -147,3 +149,115 @@ def unlock_admin(admin_id):
...
+                    'error', current_user.username,
+                    'The following error occurred in twofactor_disable: {0}'.format(str(e)))
+                raise GenericError('The administrator could not be updated')
+            return dict(enabled=admin.otp_active)
+        elif status.lower() == "true":
+            raise GenericError("Cannot enable 2 factor from this route - see docs")
+    raise GenericError("An Invalid parameter was supplied")
+
+
+@apiv1.route('/admins/<int:admin_id>/2factor/qrcode', methods=['GET'])
```
Comment (mprahl) [inline postmaster/apiv1/admins.py]: Can we change the route to "twofactor"?

File: postmaster/apiv1/admins.py [hunk 16]
```diff
@@ -147,3 +149,115 @@ def unlock_admin(admin_id):
...
+    stream = StringIO()
+    url.svg(stream, scale=5)
+    return stream.getvalue().encode('utf-8'), 200, {
+        'Content-Type': 'image/svg+xml',
+        'Cache-Control': 'no-cache, no-store, must-revalidate',
+        'Pragma': 'no-cache',
+        'Expires': '0'}
+
+
+@apiv1.route('/admins/<int:admin_id>/2factor/verify', methods=['POST'])
```
Comment (mprahl) [inline postmaster/apiv1/admins.py]: Can we change the route to "twofactor"?

File: tests/api/test_api_functions.py [hunk 17]
```diff
@@ -328,6 +329,96 @@ def test_admins_unlock_not_found(self, loggedin_client):
...
+        token = onetimepass.get_totp(secret)
+        assert test_admin.verify_totp(token)
+        rv = loggedin_client.post("/api/v1/admins/{0}/2factor/verify".format(test_admin.id), data=json.dumps({"code": token}))
+        try:
+            json.loads(rv.data)
+        except:
+            assert False, "Not json"
+        assert "Success" in rv.data
+        assert rv.status_code == 200
+        assert "Success" in rv.data
```
Comment (mprahl) [inline tests/api/test_api_functions.py]: Duplicate line. Please remove.

**Comment by mprahl** [comment]
@thatarchguy nice work. I just have some simple comments you need to address and then we'll merge this.

**Comment by mprahl** [comment]
@thatarchguy also please add to the Change Log.

File: postmaster/static/css/styles.css (hunk 8 above)
Comment (thatarchguy) [inline postmaster/static/css/styles.css]: oops, that was my editor automatically fixing that stuff

File: postmaster/utils.py (hunk 11 above)
Comment (thatarchguy) [inline postmaster/utils.py]: It's written the way it is so we can return messages based on if they even have otp_active or if their token was correct.

File: postmaster/utils.py (hunk 12 above)
Comment (thatarchguy) [inline postmaster/utils.py]: It's written the way it is so we can return messages based on if they even have otp_active or if their token was correct.

File: postmaster/apiv1/admins.py (hunk 2 above)
Comment (mprahl) [inline postmaster/apiv1/admins.py]: @thatarchguy can you take a look at this?

**Comment by mprahl** [comment]
@thatarchguy LGTM!

Next time, don't merge master into your branch, do:
```
git checkout master
git pull
git checkout feature-branch-name
git rebase master
```
This stacks your new commits on top of the latest master.
